src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs

```diff
@@ -66,30 +71,47 @@ private void CaptureStackTrace(int skipFrames, bool fNeedFileInfo, Exception? e)
             StackF.InitializeSourceInfo(fNeedFileInfo, e);
 
             _numOfFrames = StackF.GetNumberOfFrames();
+            _stackFrameHelper = StackF;
 
             if (_methodsToSkip > _numOfFrames)
                 _methodsToSkip = _numOfFrames;
 
-            if (_numOfFrames != 0)
+            _fNeedFileInfo = fNeedFileInfo;
+            _isForException = e != null;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="StackFrame"/> array for this StackTrace, possibly creating if needed.
+        /// </summary>
+        /// <returns></returns>
+        private StackFrame[] GetFramesCore()
+        {
+            if (_numOfFrames != 0 && _stackFrames == null)
             {
-                _stackFrames = new StackFrame[_numOfFrames];
+                Debug.Assert(_stackFrameHelper != null);
+
+                StackFrame[] stackFrames = new StackFrame[_numOfFrames];
 
                 for (int i = 0; i < _numOfFrames; i++)
                 {
-                    _stackFrames[i] = new StackFrame(StackF, i, fNeedFileInfo);
+                    stackFrames[i] = new StackFrame(_stackFrameHelper, i, _fNeedFileInfo);
```

Comment:
I suspect that deferring creation here will have issues with unloadable methods, since no reference to the `MethodBase` is held until this point.

Reply:
Yes, this is a problem

Reply:
I've added an array to store `MethodBase` references into the eager part of StackTrace construction for now.